src/main/scala/RocketChip.scala

```diff
@@ -225,25 +230,29 @@ class Uncore(implicit val p: Parameters) extends Module
   io.mem_axi <> outmemsys.io.mem_axi
   io.mem_ahb <> outmemsys.io.mem_ahb
   io.mem_tl  <> outmemsys.io.mem_tl
-
-  def connectExternalMMIO(ext: ClientUncachedTileLinkIO)(implicit p: Parameters) {
-    val mmio_axi = p(NExtMMIOAXIChannels)
-    val mmio_ahb = p(NExtMMIOAHBChannels)
-    val mmio_tl  = p(NExtMMIOTLChannels)
-
-    require (mmio_axi + mmio_ahb + mmio_tl <= 1)
-    if (mmio_ahb == 1) {
-      val ahb = Module(new AHBBridge(true)) // with atomics
-      io.mmio_ahb.head <> ahb.io.ahb
-      ahb.io.tl <> ext
-    } else if (mmio_tl == 1) {
-      TopUtils.connectTilelink(io.mmio_tl.head, ext)
-    } else {
-      val mmioEndpoint = mmio_axi match {
-        case 0 => Module(new NastiErrorSlave).io
-        case 1 => io.mmio_axi.head
+  outmemsys.io.bus_axi <> io.bus_axi
+
+  def connectExternalMMIO(ports: Seq[ClientUncachedTileLinkIO])(implicit p: Parameters) {
+    val mmio_axi_start = 0
+    val mmio_axi_end   = mmio_axi_start + p(NExtMMIOAXIChannels)
+    val mmio_ahb_start = mmio_axi_end
+    val mmio_ahb_end   = mmio_ahb_start + p(NExtMMIOAHBChannels)
+    val mmio_tl_start  = mmio_ahb_end
+    val mmio_tl_end    = mmio_tl_start  + p(NExtMMIOTLChannels)
+    require (mmio_tl_end == ports.size)
+    
+    for (i <- 0 until ports.size) {
+      if (mmio_axi_start <= i && i < mmio_axi_end) {
+        TopUtils.connectTilelinkNasti(io.mmio_axi(i-mmio_axi_start), ports(i))
+      } else if (mmio_ahb_start <= i && i < mmio_ahb_end) {
+        val ahbBridge = Module(new AHBBridge(true))
+        io.mmio_ahb(i-mmio_ahb_start) <> ahbBridge.io.ahb
+        ahbBridge.io.tl <> ports(i)
+      } else if (mmio_tl_start <= i && i < mmio_tl_end) {
+        TopUtils.connectTilelink(io.mmio_tl(i-mmio_tl_start), ports(i))
+      } else {
+        TopUtils.connectTilelinkNasti(Module(new NastiErrorSlave).io, ports(i))
```

Comment:
This part shouldn't be necessary. The require above already makes sure that the number of interconnect ports matches the number of external MMIOs.